### PR TITLE
Add Repository::commit_new method with permissive parent param

### PR DESCRIPTION
**Pull request by chetaldrich**
The existing method doesn't accept owned commits as parameters in
Repository::commit. This new method allows for both owned and unowned
versions of the slice, which will simplify passing in Vectors
containing owned values as well as retain backwards compatibility in
_most_ cases (unfortunately, an empty slice still needs a type specified if passed).

Fixes https://github.com/rust-lang/git2-rs/issues/140, which specifies the issue in more detail.

### CODE REVIEW

File: src/repo.rs [hunk 1]
```diff
@@ -1231,6 +1231,56 @@ impl Repository {
     /// current branch and make it point to this commit. If the reference
     /// doesn't exist yet, it will be created. If it does exist, the first
     /// parent must be the tip of this branch.
+    pub fn new_commit<'a, C: AsRef<Commit<'a>>>(
```
Comment (extrawurst) [inline src/repo.rs:1234]: can we add this `&[] as &[Commit<'_>]` for the empty list case to the documentation as this will be a little unintuitive 🙈

**Comment by extrawurst** [comment]
I personally would prefer `commit_new` but thats a nitpicky

File: src/repo.rs (hunk 1 above)
Comment (chetaldrich) [reply src/repo.rs:1234]: Sure, I'll add that in. It's really too bad the default type parameter fallback feature didn't make it into the language since this issue was created.

**Comment by chetaldrich** [comment]
I can switch over to commit_new since I'll be making that other small documentation change anyway.

File: src/repo.rs (hunk 1 above)
Comment (alexcrichton) [reply src/repo.rs:1234]: I think this may have actually been one of the issues which led me to the original signature, if this otherwise requires type annotations then I don't think that this is the best way to shape this API. Another option would be perhaps a builder-like construct but that's pretty wordy for just the type inference issue here.

Comment (chetaldrich) [reply src/repo.rs:1234]: So there are two options here: 

1. We can stick with this, which allows existing calls that contain values to be the same (appears to be the majority of calls, based on some seen usage of this method in the wild and in the tests), but does have this edge case that Rust doesn't handle very gracefully, while supporting owned values in the slice/vector parameter. 
2. We could make a hard switch to using owned values, so the signature would be `&[Commit]` in that case. We lose compatibility with existing calls, but this edge case type inference issue goes away. 

I personally think that supporting owned values makes sense here, thus the reason for the PR, but would like to support the existing usage where possible. This makes the former seem like a good compromise to me since it only comes up in this one case, but I could go either way. 

Thoughts?

**Comment by chetaldrich** [comment]
I've updated the PR to incorporate the suggestions of @extrawurst, but if we decide that the type inference issue makes us want to go the `&[Commit]` route I'm happy to make the needed changes.

**Comment by alexcrichton** [comment]
Personally I still feel like `&[&Commit<'_>],` is the best for now since it doesn't require type annotations, nor does it require ownership of the commits themselves.

**Comment by chetaldrich** [comment]
I suppose if you're committing to that then we should also close the issue, then. I'll close this.
